### ISSUE TITLE
Styles: show a hierarchy of styles clearly (theme, user, global unspecific, global specific, local, etc)

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#49278]
### When no styles are detected
<img width="1027" alt="No styles detected" src="https://github.com/WordPress/gutenberg/assets/846565/ed09eae8-cf49-4f13-9cb7-a2ee25e11c63">

When styles cannot be detected (from a style provider like global styles / theme.json) no controls are displayed. This eliminates the confusion between undetectable inherited styles (e.g. custom css, browser styles) and empty/unset controls. See #43082 for additional context. 

### Working globally
<img width="1027" alt="Working globally" src="https://github.com/WordPress/gutenberg/assets/846565/b14e4c7f-cf51-4405-87e1-9dc332e3c6a5">

Respecting the convention above, panels are empty until a value is set (either in the UI, or via theme.json).

### Working locally
When working locally there are three states to consider:

**Inherited values**
<img width="1027" alt="Working locally" src="https://github.com/WordPress/gutenberg/assets/846565/551ac2cb-b7eb-4d7b-a291-f206fe435eaf">

If a value is inherited (e.g. from global styles) the control displays that value, and the label has a purple accent denoting the global context. Hovering/focussing the label reveals the source in a tooltip. The control cannot be removed because if there's no local setting, the inherited value will be used.

**Local overrides**
<img width="1027" alt="Local overrides" src="https://github.com/WordPress/gutenberg/assets/846565/1a248a6a-6151-4667-8262-bb972f74e324">

If an inherited value is changed then a blue dot is appended to the label. Clicking the dot reveals a menu where the user can reset to the global default, or push the value upstream. Some controls (like background image, see below) may need an additional action to 'unset' the value locally.

**Local styles**
Styles can be added to a block in the local context that have no inherited value. The manifestation of this in the UI is essentially the same as the "Working globally" case above:

<img width="1027" alt="Working globally" src="https://github.com/WordPress/gutenberg/assets/846565/b14e4c7f-cf51-4405-87e1-9dc332e3c6a5">

With one caveat: If a global style for this property is later added, the local control would then exhibit the blue dot. 

### Examples
In addition to the block-spacing example above, here are a couple others that have come up in other issues recently. In each case the edit context is local to a document:

**Lightbox**
<img width="888" alt="Lightbox" src="https://github.com/WordPress/gutenberg/assets/846565/ccd18230-9bbd-4cf6-9250-a162c5027abf">

**Background image**
<img width="1226" alt="Background image" src="https://github.com/WordPress/gutenberg/assets/846565/3b85ea1e-26af-46a6-8809-df61a3b81d06">

### Breakdown

This doesn't all need to be implemented at once. The work could be broken down into the following steps:

- [ ] https://github.com/WordPress/gutenberg/issues/37752
- [ ] If local styles override global then indicate that change with a dot which serves as a Reset affordance.
- [ ] Highlight where those styles are coming from.
- [ ] Implement make default/push to global functionality.

### Important note

The purple accents used to indicate inherited styles in the mockups above could be a bit heavy-handed in practise. We should continue to explore other design solutions for this.

<details><summary>Original issue</summary>Knowing what style is affecting something on a theme is difficult (is it a global element? is it coming from a block type? is it local to the template?). This has shown up in a few areas already, from Custom CSS https://github.com/WordPress/gutenberg/issues/47946 to the general experience of conflicting styles https://github.com/WordPress/gutenberg/issues/45086 https://github.com/WordPress/gutenberg/issues/44931. To improve this, we should find a way to show a hierarchy of styles clearly (theme, user, global unspecific, global specific, local, etc). While building this, we should also consider ways to push or transfer something to the right level when working through a design, which relates to https://github.com/WordPress/gutenberg/issues/44931

![Image](https://user-images.githubusercontent.com/1204802/238591076-37eb7040-590b-496c-810b-efb46180ca83.png)

In most cases understanding this style hierarchy is not crucial, as you can accomplish most of what you need to without the visualization. You go to global styles to style site wide, you edit pages to style locally. Nevertheless for more complex sessions, it can be useful to surface, but due to the rarity of needing to surface this, the footprint and UI weight of any visualization feature needs to be small. 

Here's a take that makes the design tool _label_ clickable to reveal the inheritance:

![Image](https://user-images.githubusercontent.com/1204802/238591835-7af6e800-3059-4803-b5a9-da815eaf6aac.png)

* Click a label to see the inheritance
* If the popover has a link, that link should serve as a deep-link to the origin of the inheritance

For this concept to work, the Color flyout has received its own label.

Thoughts?</details>

**Comment by jasmussen** [comment]
As an example of style inheritance reflected in a few ways:

* If no font is set in global styles, it inherits Times New Roman from the browser stylesheet (though I think we may have overridden that to system fonts)
* If "Crimson Pro" is set as a global font, then a group inherits that. But a group can also further set its own font inside, such as DM Sans. 
* Subsequent nested groups could keep at this, creating a deep inheritance hierarchy, i.e. Group ← Group ← Group ← Global Styles ← Browse styles.

The added nuance here is that in the block editor, we are talking not just about CSS inheritance, but also about _global styles inheritance_, such as inherited content/wide widths. And even further, you might set a block default in Global Styles > Blocks, and then override it locally in a post. In the future you might even set _ref'ed_ colors (Set the background color of this button to the same color as the global text color). 

Finally, there's theme inheritance. You might activate a block theme that has a default root padding. You might then create your own style variation of that, which has its own root padding. It needs here to be visualized that your style variation is replacing root padding values that exist in theme.json.

Sharing these examples as context for a design that tackles this, that ideally such a design can tackle these, while still being minimalistic, and perhap ideally with a way to navigate to each origin.

Related:

* #43082
* #37752

**Comment by jasmussen** [comment]
I ended up creating https://github.com/WordPress/gutenberg/issues/50659 instead of updating this one. I'm happy to close the other one and use this one? Let me know.

**Comment by annezazu** [comment]
It might seem silly but I've linked to this one a fair amount. Mind closing yours and sharing the designs here?

**Comment by jasmussen** [comment]
Not at all, the other one was created by mistake! Updated this one. Thanks Anne!

**Comment by annezazu** [comment]
[Passing along feedback](https://make.wordpress.org/test/2023/05/18/fse-program-testing-call-23-rapid-revamp/#comment-3069) from the FSE Outreach Program's Rapid Revamp call for testing that underscores how painful this currently is:

>when I changed the size of the header type h2, I hoped that the titles in the Query Block changed too, but it didn’t change for me. Then, I think that the Post Title block into the Query Block had a custom size, but it had the default size. I reviewed the Blocks in Styles section, but it had the default size too. Same issue with the excerpt, when I change the Text size, don’t change in the excerpt. Also occur with Appearance and Line Height. The Font Family works fine, previewing the changes. Inspecting the front end, I view that get defaults values in the Query Block. For example, in the title have .wp-block-query h2 style. It’s confused for me. I think that should be got the default h2 styles, except if it is customized.

**Comment by jameskoster** [comment]
One thing that has been historically challenging to communicate is the presence of _default_ values. The label treatment suggested here effectively handles this in a way that can be applied consistently across all controls. With that said there may be an argument to increase the prominence slightly by making use of the purple 'sync' color to help indicate there's a connection beyond the local scope. 

Here's a quick mockup:

<img width="1082" alt="Inherited + local styles" src="https://github.com/WordPress/gutenberg/assets/846565/5b9ffc3b-0778-4198-baf2-9a08b9e6dde3">

**Comment by jasmussen** [comment]
> With that said there may be an argument to increase the prominence slightly by making use of the purple 'sync' color to help indicate there's a connection beyond the local scope.

I see where you're coming from, but I worry that when you see a full editor with a canvas, a top toolbar, and a taller inspector, and end up with a bunch of purple splotches, it'll inevitably add some weight to what is, truly, a power-user feature. In that light, I would not highlight items inside an itemgroup — I'd still keep it label-specific (you could still click a color item, and we could add a label you can hover in the flyout). 

I wonder, can we add some purple on hover _only_? I.e can we still have a more muted treatment of the resting state? (Perhaps it's still my recent AFK that's in my blood but coming back to this I'm even thinking that the dashed underline is too prominent here 🤔 )

**Comment by jameskoster** [comment]
> I wonder, can we add some purple on hover only? I.e can we still have a more muted treatment of the resting state?

Yup, definitely something to explore, if we think the general heuristic makes sense. In that regard, we may consider the opposite mechanic: instead of highlighting controls that are set to their default values, highlight ones that are _not_, perhaps with a blue dot or something similar. This could provide a direct pathway to resetting individual values too:

<img width="1191" alt="Screenshot 2023-07-31 at 10 59 13" src="https://github.com/WordPress/gutenberg/assets/846565/1e1f56d1-75a4-40da-95e2-16052833a543">

**Comment by jasmussen** [comment]
> highlight ones that are not, perhaps with a blue dot or something similar. This could provide a direct pathway to resetting individual values too:

Oh that's interesting. So anything that's inherited, regardless of origin, shows programmatically detected values, but anything else have a dot. How would the slider in your example look, in either state?

* Inherited from global styles
* Locally overridden

**Comment by jameskoster** [comment]
These are the states I think we need to account for:

* Inherited styles that _cannot_ be detected programatically, e.g. custom css or browser styles (this is basically #43082).
  * Local styles that override the above. An example of this would be styling a block in the global styles panel. Another would be adding padding to a Group while editing a post when there is no padding assigned to Group blocks in global styles.
* Inherited styles that _can_ be detected. 
  * Local styles that overrides the above. E.g. locally styling a block while editing a post.

The second point(s) is the one that is most important to communicate to users.

Here's a comparison of the approaches above:

## Highlighting local overrides

![blue-dot](https://github.com/WordPress/gutenberg/assets/846565/44f3b8fd-95d6-4d3f-a075-b4b24a99aeba)


## Highlighting inherited defaults from style provider

![purple](https://github.com/WordPress/gutenberg/assets/846565/8a1fec39-e61a-4aa0-8714-a5d2821a38d7)

**Comment by jasmussen** [comment]
I think this might just work. @SaxonF  @richtabor any thoughts, any nuance we're missing here?

**Comment by SaxonF** [comment]
I like the dot to indicate which properties have been overridden. We can use a similar pattern when making changes to a style variation in a similar way to keynote. When someone pushes local styles to global those dots just disappear which is another good bit of feedback.

**Comment by jameskoster** [comment]
It occurred to me that we could also combine both concepts above:

<img width="1715" alt="inheritance" src="https://github.com/WordPress/gutenberg/assets/846565/074ed776-63bb-41b5-8251-c06bfb67ce89">

**Comment by jasmussen** [comment]
Promising.

The purple is only on hovering the title, right? I would avoid showing that purple in its resting state, that'll just be noisy for what is a feature that is extremely useful for developers, but also very advanced.

I think we should left-align the text in the tooltip.

**Comment by jameskoster** [comment]
> The purple is only on hovering the title, right? 

Probably, yes. With the blue dot visibly indicating overrides at a glance, inheritance details can be more subtle.

**Comment by SaxonF** [comment]
Copying response [here](https://github.com/WordPress/gutenberg/issues/43082#issuecomment-1702120445) which I think can be closed if we moved forward.

It sounds like we can break this down in to a few shippable releases:

1. Always show default values in the inspector if we know them
2. Highlight where those styles are coming from
3. If local styles override global then indicate that change with a dot (as seen https://github.com/WordPress/gutenberg/issues/49278#issuecomment-1680274727) which offers a shortcut to resetting and pushing changes to global.
4. Implement make default/push to global functionality

Is there any other points of clarification we need before moving to dev? I think we're in a decent spot and can work through edge cases as they arise .

**Comment by youknowriad** [comment]
One downside of the design proposed above is that if the default is value "3" and I want to use the value "3" for my block style/setting/whatever explicitly. I mean I don't want that value to change if I switch themes or styles variations. How do I do that? I just "focus" the input, I change back and forth. It's a bit weird.

**Comment by jameskoster** [comment]
It's a _bit_ weird, but I don't think it's too bad. You have to perform a similar action if you want to replace a global value in Figma and it works okay:

![radius](https://github.com/WordPress/gutenberg/assets/846565/ba86680d-657c-466a-a47b-8313cba67398)

As a secondary affordance we might offer an action to detach/unlink from the inherited value in a single click. To connect a dot, such an action has already been suggested for the box shadow control for customising presets: https://github.com/WordPress/gutenberg/issues/44651.

**Comment by youknowriad** [comment]
Works for me, just wanted to mention it so it's something we're aware of and making a conscious decision about it.

**Comment by jameskoster** [comment]
Testing the proposed design against the lightbox control, and trying a single-click detach affordance:

<img width="2214" alt="lightbox" src="https://github.com/WordPress/gutenberg/assets/846565/f696378e-9905-4810-aa8f-b6c70f65c011">

I'm not entirely sure we need it — in most cases I suspect it's probably easier to just tweak the control. Still, it's a good one to have up the sleeve, but we'd probably need more intuitive language than 'detach'.

**Comment by youknowriad** [comment]
@jameskoster I also suspect that we have to keep the button separate from the label of the control for a11y reasons.

**Comment by jameskoster** [comment]
I updated the OP to include detailed versions of the designs as discussed throughout the issue. We can (and should) continue to discuss, but hopefully it's ready to be picked up for development.

**Comment by michalczaplinski** [comment]
I've created a very first and very small attempt to tackle the issue of showing whether a particular value is inherited or not and to validate the approach. It only works for the text color at the moment:

https://github.com/WordPress/gutenberg/pull/55952

**Comment by afercia** [comment]
@SergeyBiryukov, @carolinan and myself quickly looked into this issue this morning. Chimin in to quickly note that any informration about the 'inherited' style/setting must be provided in an accessible way.

Looking at this mockup, it seems less than ideal:

<img width="2214" alt="inherited feedback" src="https://github.com/WordPress/gutenberg/assets/1682452/00e8c1ba-5c3d-4167-9bf5-39532cfd91a3">

- Purple color: it would be only a color change, which is not sufficient to provide the information in an accessible way.
- The icon is a little better, at least it's a clear _visual_ indication. Not sure the 'unlink' icon is appropriate though. Still not ideal though, because icon-only controls come with inherent usability and accessibility issues and should be avoided.
- Tooltips should not be used for descriptions. The original intent of tooltips is to visually expose the accessible name of an icon-only control. Tooltips usage should be limited to that purpose.
- The blue dot would be an insufficient indication. We already tried a blue dot in the Templates list a while ago and removed it in favor of visible text.

Ideally, for better clarity for all users and better accessibility, visible text is always preferable. It could be something along these lines:

`Inherited from global styles/settings` + an UI control to disable inheritance.

**Comment by jasmussen** [comment]
As an update, #37752 is in a place where it can serve as a good first step.

**Comment by mtias** [comment]
With block bindings using the purple signal for conveying "connection", do we need to revise the treatment or color here?

**Comment by jameskoster** [comment]
Yeah it would probably be good to document exactly what purple is supposed to signify so we can apply consistently moving forwards.

Iirc it began as a way to signify 'global' elements which is why template parts and synced blocks exhibit purple accents. But in the "connection" paradigm I suppose it signifies something more like "an aspect of this thing is affected by another thing". E.g. a Heading block contents being populated by the post title. So it's not always _global_, but I guess it still indicates "outside influence" which is kind of similar. 

In that sense, purple here might make sense because it's utilised when the styles are being inherited from outside of the immediate context.

**Comment by mtias** [comment]
Maybe, though it's conceivable that we might have bindings for style attributes, so a color comes from a custom field, etc. In which case differentiating the hierarchy from how the attribute is assigned may still be useful? The other thing is that the purple signifies an explicit connection — _this_ attribute is connected to _this_ field; this block container is connected to this block object. While here it's about more passive inheritance.

**Comment by jameskoster** [comment]
Yeah I suppose this is more of a one-way 'connection'. 

I think the design can work with a different color, or a different treatment. As long as there's a way to differentiate local styles from inherited ones.

<img width="364" height="222" alt="Image" src="https://github.com/user-attachments/assets/42029c1e-716d-438d-9f98-b89229068a6d" />

**Comment by aaronrobertshaw** [comment]
A new attempt at implementing some of this is now tracked at #77595. It covers: step 1 of the breakdown proposed here (displaying inherited values) plus parts of step 2 (the reset affordance). Steps 3 (source attribution via tooltip) and 4 (push-to-global) are explicitly deferred to follow-up issues.

**Comment by annezazu** [comment]
> If an inherited value is changed then a blue dot is appended to the label. Clicking the dot reveals a menu where the user can reset to the global default, or push the value upstream. Some controls (like background image, see below) may need an additional action to 'unset' the value locally.

@WordPress/gutenberg-design I want to talk about this part of the design. [In testing the PR Aaron](https://github.com/WordPress/gutenberg/pull/77894) is working on to implement this (🌟 ), having the option to push globally feels like surfacing a lot of complexity and risk, for lack of a better word (risk in someone accidentally pushing something global they didn't want). When we think about what we show by default to users, like what default typography controls are available, it's always been a balancing out. This feels a touch out of balance to me but curious to hear more perspectives.

**Comment by annezazu** [comment]
Here's an idea to explore that accomplishes a few things: 

- Still gives the option to apply globally more granularly. 
- Brings the complexity of the option into the Advanced section (where I think it belongs). 
- Connects it to the current flow for applying all styles globally.
- Adds what feels like "good friction" to the apply globally flow so folks know exactly what they are doing (something I haven't loved about the current flow).

I know I'm digging into details here but this feels particularly important as any local change that's overridden will otherwise show this option and my concern is more folks will mistakenly set things globally without understanding the impact:

https://github.com/user-attachments/assets/71f3f016-5089-40de-80f0-a1900ee10bfe

**Comment by fcoveram** [comment]
Nice ✨ I like the above and how it relies on existing patterns. It makes totally sense to me.

It would be nice to create a UX pattern for saving/apply multiple changes. I was playing with an idea time ago using DataViews for an extended version of this modal where "before" and "after" are listed.

<img width="2452" height="1648" alt="Image" src="https://github.com/user-attachments/assets/9598d4da-79be-4931-a02a-aac833608f71" />

Not a blocker but sharing the idea to spark new thoughts.

**Comment by aaronrobertshaw** [comment]
I like the new approach discussed here 👍 

The PR introducing the local override indicator etc over in https://github.com/WordPress/gutenberg/pull/77894 has been updated to remove the push individual style to global styles feature. I'm happy to help coordinate or pick up this new approach to pushing to global styles if you'd like @annezazu

**Comment by annezazu** [comment]
Please do pick it up! I think it's also okay if it ends up not making it for 7.1 but of course I'd love to see it. I'd rather see us for sure land displaying inherited styles as that's a larger problem and a more common use case than pushing granular global styles.

**Comment by aaronrobertshaw** [comment]
I've made a start on a combination of https://github.com/WordPress/gutenberg/issues/49278#issuecomment-4806213008 and https://github.com/WordPress/gutenberg/issues/49278#issuecomment-4808114939 over in https://github.com/WordPress/gutenberg/pull/79839. 

At present it displays the proposed modal with a DataViews table for the local block instance styles that can be pushed to Global Styles. To limit some of the repetitiveness of listing styles things like spacing and border styles are condensed to shorthand representations in the "current" and "new" columns.

**Comment by aaronrobertshaw** [comment]
A new iteration issue to continue iterating on the display of global styles in the block inspector for WP7.2 has been created in:
- https://github.com/WordPress/gutenberg/issues/82520
